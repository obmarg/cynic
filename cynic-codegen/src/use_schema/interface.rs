use quote::{quote, ToTokens, TokenStreamExt};

use crate::schema::types::InterfaceType;

use super::fields::FieldOutput;

pub struct InterfaceOutput<'a> {
    iface: InterfaceType<'a>,
    marker_ident: proc_macro2::Ident,
}

impl<'a> InterfaceOutput<'a> {
    pub fn new(iface: InterfaceType<'a>) -> Self {
        InterfaceOutput {
            marker_ident: iface.marker_ident().to_rust_ident(),
            iface,
        }
    }

    pub fn append_fields(&self, field_module: &mut proc_macro2::TokenStream) {
        if !self.iface.fields.is_empty() {
            let field_module_ident = self.iface.field_module().ident();
            let fields = self.iface.fields.iter().map(|f| FieldOutput {
                field: f,
                parent_marker: &self.marker_ident,
            });
            field_module.append_all(quote! {
                pub mod #field_module_ident {
                    #(#fields)*
                }
            });
        }
    }
}

impl ToTokens for InterfaceOutput<'_> {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        let marker_ident = &self.marker_ident;
        tokens.append_all(quote! {
            pub struct #marker_ident;
        });
    }
}
