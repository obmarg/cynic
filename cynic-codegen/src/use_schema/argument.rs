use {
    quote::{quote, ToTokens, TokenStreamExt},
    syn::parse_quote,
};

use crate::schema::types::InputValue;

pub struct ArgumentOutput<'a> {
    argument: &'a InputValue<'a>,

    // Marker for the field or directive this is contained within
    container_marker: &'a proc_macro2::Ident,

    kind: ArgumentKind,
}

enum ArgumentKind {
    Directive,
    Field,
}

impl<'a> ArgumentOutput<'a> {
    pub fn field_argument(
        argument: &'a InputValue<'a>,
        container_marker: &'a proc_macro2::Ident,
    ) -> Self {
        ArgumentOutput {
            argument,
            container_marker,
            kind: ArgumentKind::Field,
        }
    }

    pub fn directive_argument(
        argument: &'a InputValue<'a>,
        container_marker: &'a proc_macro2::Ident,
    ) -> Self {
        ArgumentOutput {
            argument,
            container_marker,
            kind: ArgumentKind::Directive,
        }
    }
}

impl ToTokens for ArgumentOutput<'_> {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        let name = proc_macro2::Literal::string(self.argument.name.as_str());
        let argument_ident = self.argument.marker_ident().to_rust_ident();
        let field_marker = self.container_marker;

        let prefix = match self.kind {
            ArgumentKind::Directive => parse_quote! { super },
            ArgumentKind::Field => parse_quote! { super::super::super },
        };

        let schema_type = self.argument.value_type.marker_type().to_path(&prefix);

        tokens.append_all(quote! {
            pub struct #argument_ident;

            impl cynic::schema::HasArgument<#argument_ident> for super::#field_marker {
                type ArgumentType = #schema_type;

                const NAME: &'static ::core::primitive::str = #name;
            }
        })
    }
}
