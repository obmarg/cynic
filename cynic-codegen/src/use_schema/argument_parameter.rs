use crate::{Errors, FieldType, Ident};
use proc_macro2::TokenStream;

/// A parmeter to a schema function that represents a GraphQL parameter.
pub struct ArgumentParameter {
    name: Ident,
    param_type: ArgumentParameterType,
}

impl ArgumentParameter {
    pub fn new(name: Ident, argument_type: FieldType) -> ArgumentParameter {
        ArgumentParameter {
            name,
            param_type: ArgumentParameterType::from_type(argument_type),
        }
    }

    pub fn to_tokens(&self, path_to_markers: syn::Path) -> Result<TokenStream, Errors> {
        use quote::quote;

        let name = &self.name;
        let type_tokens = self.param_type.to_tokens(path_to_markers)?;

        Ok(quote! { #name: #type_tokens })
    }
}

pub struct ArgumentParameterType {
    argument_type: FieldType,
}

impl ArgumentParameterType {
    pub fn from_type(argument_type: FieldType) -> ArgumentParameterType {
        ArgumentParameterType { argument_type }
    }

    pub fn to_tokens(&self, path_to_markers: syn::Path) -> Result<TokenStream, Errors> {
        use quote::quote;

        let type_lock = self.argument_type.as_type_lock(&path_to_markers);
        let wrapper_path = self.argument_type.wrapper_path()?;

        Ok(quote! { impl ::cynic::InputType<#type_lock, #wrapper_path> })
    }
}
