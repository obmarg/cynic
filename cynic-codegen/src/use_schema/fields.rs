use {
    quote::{quote, ToTokens, TokenStreamExt},
    syn::parse_quote,
};

use crate::schema::types::{Field, InputValue};

pub struct FieldOutput<'a> {
    pub(super) field: &'a Field<'a>,
    pub(super) parent_marker: &'a proc_macro2::Ident,
}

struct ArgumentOutput<'a> {
    argument: &'a InputValue<'a>,
    field_marker: &'a proc_macro2::Ident,
}

impl ToTokens for FieldOutput<'_> {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        let parent_marker = self.parent_marker;
        let field_marker = &self.field.marker_ident().to_rust_ident();
        let field_name_literal = proc_macro2::Literal::string(self.field.name.as_str());

        let field_type_marker = self
            .field
            .field_type
            .marker_type()
            .to_path(&parse_quote! { super::super });

        tokens.append_all(quote! {
            pub struct #field_marker;

            impl cynic::schema::Field for #field_marker{
                type Type = #field_type_marker;

                const NAME: &'static str = #field_name_literal;
            }

            impl cynic::schema::HasField<#field_marker> for super::super::#parent_marker {
                type Type = #field_type_marker;
            }
        });

        if !self.field.arguments.is_empty() {
            let argument_module = self.field.argument_module().ident();
            let arguments = self.field.arguments.iter().map(|argument| ArgumentOutput {
                argument,
                field_marker,
            });

            tokens.append_all(quote! {
                pub mod #argument_module {
                    #(#arguments)*
                }
            });
        }
    }
}

impl ToTokens for ArgumentOutput<'_> {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        let name = proc_macro2::Literal::string(self.argument.name.as_str());
        let argument_ident = self.argument.marker_ident().to_rust_ident();
        let field_marker = self.field_marker;

        let schema_type = self
            .argument
            .value_type
            .marker_type()
            .to_path(&parse_quote! { super::super::super  });

        tokens.append_all(quote! {
            pub struct #argument_ident;

            impl cynic::schema::HasArgument<#argument_ident> for super::#field_marker {
                type ArgumentType = #schema_type;

                const NAME: &'static str = #name;
            }
        })
    }
}
