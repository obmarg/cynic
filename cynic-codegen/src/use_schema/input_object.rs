use quote::{quote, ToTokens, TokenStreamExt};
use syn::parse_quote;

use crate::schema::types::{InputObjectType, InputValue};

pub struct InputObjectOutput<'a> {
    object: InputObjectType<'a>,
}

struct FieldOutput<'a> {
    field: &'a InputValue<'a>,
    object_marker: &'a proc_macro2::Ident,
}

impl<'a> InputObjectOutput<'a> {
    pub fn new(object: InputObjectType<'a>) -> Self {
        InputObjectOutput { object }
    }
}

impl ToTokens for InputObjectOutput<'_> {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        let object_marker = proc_macro2::Ident::from(self.object.marker_ident());
        tokens.append_all(quote! {
            pub struct #object_marker;

            impl ::cynic::schema::InputObjectMarker for #object_marker {}
        });

        if !self.object.fields.is_empty() {
            let field_module = self.object.field_module().ident();
            let fields = self.object.fields.iter().map(|f| FieldOutput {
                field: f,
                object_marker: &object_marker,
            });
            tokens.append_all(quote! {
                pub mod #field_module {
                    #(#fields)*
                }
            });
        }
    }
}

impl ToTokens for FieldOutput<'_> {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        let object_marker = self.object_marker;
        let field_marker = &proc_macro2::Ident::from(self.field.marker_ident());
        let field_name_literal = proc_macro2::Literal::string(self.field.name.as_str());

        let field_type_marker = self
            .field
            .value_type
            .marker_type()
            .to_path(&parse_quote! { super });

        tokens.append_all(quote! {
            pub struct #field_marker;

            impl ::cynic::schema::Field for #field_marker{
                type Type = #field_type_marker;

                const NAME: &'static str = #field_name_literal;
            }

            impl ::cynic::schema::HasInputField<#field_marker, #field_type_marker> for super::#object_marker {
            }
        });
    }
}
