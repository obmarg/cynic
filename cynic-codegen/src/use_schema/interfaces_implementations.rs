use proc_macro2::TokenStream;

use crate::{schema, Ident};

/// We generate an InterfaceImplementation for each type that implements interface.
///
/// These are output as `HasSubtype` implementations.
#[derive(Debug)]
pub struct InterfacesImplementations {
    pub implementor: Ident,
    pub interfaces: Vec<Ident>,
}

impl InterfacesImplementations {
    pub fn from_interface(iface: &schema::InterfaceType) -> Self {
        let ident = Ident::for_type(&iface.name);

        Self {
            implementor: ident.clone(),
            interfaces: vec![ident],
        }
    }

    pub fn from_object(obj: &schema::ObjectType) -> Option<Self> {
        if obj.implements_interfaces.is_empty() {
            return None;
        }

        Some(Self {
            implementor: Ident::for_type(&obj.name),
            interfaces: obj
                .implements_interfaces
                .iter()
                .map(Ident::for_type)
                .collect(),
        })
    }
}

impl quote::ToTokens for InterfacesImplementations {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        use quote::{quote, TokenStreamExt};

        let implementor = &self.implementor;
        let interfaces = &self.interfaces;

        tokens.append_all(quote! {
            #(
                impl ::cynic::selection_set::HasSubtype<#implementor> for #interfaces {}
            )*
        });
    }
}
