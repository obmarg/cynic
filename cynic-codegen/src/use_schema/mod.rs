mod fields;
mod input_object;
mod interface;
mod named_type;
mod object;
mod params;
mod schema_roots;
mod subtype_markers;

pub use params::UseSchemaParams;

use proc_macro2::TokenStream;
use quote::{quote, ToTokens};

use crate::{
    error::Errors,
    idents::Ident,
    schema::{types::Type, Schema},
};

use self::{
    input_object::InputObjectOutput, interface::InterfaceOutput, named_type::NamedType,
    object::ObjectOutput, subtype_markers::SubtypeMarkers,
};

pub fn use_schema(input: UseSchemaParams) -> Result<TokenStream, Errors> {
    use quote::TokenStreamExt;

    let document = crate::schema::load_schema(input.schema_filename)
        .map_err(|e| e.into_syn_error(proc_macro2::Span::call_site()))?;
    let schema = Schema::new(&document).validate()?;

    let mut output = TokenStream::new();

    let root_types = schema_roots::RootTypes::from_definitions(&document.definitions);
    output.append_all(quote! {
        #root_types
    });

    let mut subtype_markers = Vec::new();
    let mut named_types = Vec::new();

    for definition in schema.iter() {
        named_types.extend(NamedType::from_def(&definition));

        match definition {
            Type::Scalar(def) if !def.builtin => {
                let ident = proc_macro2::Ident::from(def.marker_ident());
                output.append_all(quote! {
                    pub struct #ident {}
                });
            }
            Type::Scalar(_) => {}
            Type::Object(def) => {
                subtype_markers.extend(SubtypeMarkers::from_object(&def));

                ObjectOutput::new(def).to_tokens(&mut output);
            }
            Type::Interface(def) => {
                subtype_markers.push(SubtypeMarkers::from_interface(&def));

                InterfaceOutput::new(def).to_tokens(&mut output);
            }
            Type::Union(def) => {
                subtype_markers.extend(SubtypeMarkers::from_union(&def));

                let ident = Ident::for_type(&def.name);
                output.append_all(quote! {
                    pub struct #ident {}
                });
            }
            Type::Enum(def) => {
                let ident = proc_macro2::Ident::from(def.marker_ident());
                output.append_all(quote! {
                    pub struct #ident {}
                });
            }
            Type::InputObject(def) => {
                InputObjectOutput::new(def).to_tokens(&mut output);
            }
        }
    }

    output.append_all(quote! {
        #(#subtype_markers)*
        #(#named_types)*

        pub type Boolean = bool;
        pub type String = std::string::String;
        pub type Float = f64;
        pub type Int = i32;
        pub type ID = ::cynic::Id;

        pub mod variable {
            use ::cynic::variables::VariableType;

            /// Used to determine the type of a given variable that
            /// appears in an argument struct.
            pub trait Variable {
                const TYPE: VariableType;
            }

            impl<T> Variable for &T where T: Variable {
                const TYPE: VariableType = T::TYPE;
            }

            impl<T> Variable for Option<T>
            where
                T: Variable
            {
                const TYPE: VariableType = VariableType::Nullable(&T::TYPE);
            }

            impl<T> Variable for Vec<T>
            where
                T: Variable,
            {
                const TYPE: VariableType = VariableType::List(&T::TYPE);
            }

            impl<T> Variable for Box<T>
            where
                T: Variable,
            {
                const TYPE: VariableType = T::TYPE;
            }

            impl<T> Variable for std::rc::Rc<T>
            where
                T: Variable,
            {
                const TYPE: VariableType = T::TYPE;
            }

            impl<T> Variable for std::sync::Arc<T>
            where
                T: Variable,
            {
                const TYPE: VariableType = T::TYPE;
            }

            impl Variable for bool {
                const TYPE: VariableType = VariableType::Named("Boolean");
            }

            impl Variable for String {
                const TYPE: VariableType = VariableType::Named("String");
            }

            impl Variable for f64 {
                const TYPE: VariableType = VariableType::Named("Float");
            }

            impl Variable for i32 {
                const TYPE: VariableType = VariableType::Named("Int");
            }

            impl Variable for ::cynic::Id {
                const TYPE: VariableType = VariableType::Named("ID");
            }
        }
    });

    Ok(output)
}
