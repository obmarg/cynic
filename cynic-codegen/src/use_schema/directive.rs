use quote::{quote, ToTokens, TokenStreamExt};

use crate::schema::types::Directive;

use super::argument::ArgumentOutput;

pub struct FieldDirectiveOutput<'a> {
    pub(super) directive: &'a Directive<'a>,
}

impl ToTokens for FieldDirectiveOutput<'_> {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        let directive_marker = self.directive.marker_ident().to_rust_ident();
        let directive_name_literal = proc_macro2::Literal::string(&self.directive.name);
        tokens.append_all(quote! {
            #[allow(non_camel_case_types)]
            pub struct #directive_marker;

            impl cynic::schema::FieldDirective for #directive_marker {
                const NAME: &'static str = #directive_name_literal;
            }
        });

        if !self.directive.arguments.is_empty() {
            let argument_module = self.directive.argument_module().ident();
            let arguments =
                self.directive.arguments.iter().map(|argument| {
                    ArgumentOutput::directive_argument(argument, &directive_marker)
                });

            tokens.append_all(quote! {
                #[allow(non_camel_case_types)]
                pub mod #argument_module {
                    #(#arguments)*
                }
            });
        }
    }
}
