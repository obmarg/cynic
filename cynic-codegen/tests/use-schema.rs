use std::{io::Write, path::PathBuf, process::Stdio};

use insta::assert_snapshot;
use rstest::rstest;

use cynic_codegen::use_schema::{use_schema, QueryDslParams};

// TODO: Rename this file after running snapshots

#[rstest(schema_file => [
    "graphql.jobs.graphql",
    "books.graphql",
    "starwars.schema.graphql"
])]
fn snapshot_use_schema(schema_file: &str) {
    let schema_path = PathBuf::from("../schemas/").join(schema_file);

    let tokens = use_schema(QueryDslParams {
        schema_filename: schema_path.to_str().unwrap().to_string(),
    })
    .unwrap();

    assert_snapshot!(format_code(format!("{}", tokens)));
}

fn format_code(input: String) -> String {
    let mut cmd = std::process::Command::new("rustfmt")
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::inherit())
        .spawn()
        .expect("failed to execute rustfmt");

    write!(cmd.stdin.as_mut().unwrap(), "{}", input).unwrap();

    std::str::from_utf8(&cmd.wait_with_output().unwrap().stdout)
        .unwrap()
        .to_owned()
}
