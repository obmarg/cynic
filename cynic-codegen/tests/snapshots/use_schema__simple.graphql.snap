---
source: cynic-codegen/tests/use-schema.rs
expression: "format_code(format!(\"{}\", tokens))"
---
impl ::cynic::schema::QueryRoot for Query {}
pub struct AnInputType;
impl ::cynic::schema::InputObjectMarker for AnInputType {}
pub mod an_input_type_fields {
    pub struct favouriteDessert;
    impl ::cynic::schema::Field for favouriteDessert {
        type Type = Option<super::Dessert>;
        const NAME: &'static str = "favouriteDessert";
    }
    impl ::cynic::schema::HasInputField<favouriteDessert, Option<super::Dessert>>
        for super::AnInputType
    {
    }
}
pub struct Dessert {}
pub struct JSON {}
impl ::cynic::schema::NamedType for JSON {
    const NAME: &'static str = "JSON";
}
pub struct MyUnionType {}
pub struct Nested;
pub mod nested_fields {
    pub struct aString;
    impl ::cynic::schema::Field for aString {
        type Type = super::String;
        const NAME: &'static str = "aString";
    }
    impl ::cynic::schema::HasField<aString> for super::Nested {
        type Type = super::String;
    }
    pub struct optString;
    impl ::cynic::schema::Field for optString {
        type Type = Option<super::String>;
        const NAME: &'static str = "optString";
    }
    impl ::cynic::schema::HasField<optString> for super::Nested {
        type Type = Option<super::String>;
    }
}
pub struct Query;
pub mod query_fields {
    pub struct testStruct;
    impl ::cynic::schema::Field for testStruct {
        type Type = Option<super::TestStruct>;
        const NAME: &'static str = "testStruct";
    }
    impl ::cynic::schema::HasField<testStruct> for super::Query {
        type Type = Option<super::TestStruct>;
    }
    pub struct myUnion;
    impl ::cynic::schema::Field for myUnion {
        type Type = Option<super::MyUnionType>;
        const NAME: &'static str = "myUnion";
    }
    impl ::cynic::schema::HasField<myUnion> for super::Query {
        type Type = Option<super::MyUnionType>;
    }
}
pub struct TestStruct;
pub mod test_struct_fields {
    pub struct fieldOne;
    impl ::cynic::schema::Field for fieldOne {
        type Type = super::String;
        const NAME: &'static str = "fieldOne";
    }
    impl ::cynic::schema::HasField<fieldOne> for super::TestStruct {
        type Type = super::String;
    }
    pub mod field_one_arguments {
        pub struct x;
        impl ::cynic::schema::HasArgument<x> for super::fieldOne {
            type ArgumentType = Option<super::super::Int>;
            const NAME: &'static str = "x";
        }
        pub struct y;
        impl ::cynic::schema::HasArgument<y> for super::fieldOne {
            type ArgumentType = Option<super::super::String>;
            const NAME: &'static str = "y";
        }
    }
    pub struct tastyCakes;
    impl ::cynic::schema::Field for tastyCakes {
        type Type = super::Dessert;
        const NAME: &'static str = "tastyCakes";
    }
    impl ::cynic::schema::HasField<tastyCakes> for super::TestStruct {
        type Type = super::Dessert;
    }
    pub mod tasty_cakes_arguments {
        pub struct first;
        impl ::cynic::schema::HasArgument<first> for super::tastyCakes {
            type ArgumentType = super::super::Dessert;
            const NAME: &'static str = "first";
        }
        pub struct second;
        impl ::cynic::schema::HasArgument<second> for super::tastyCakes {
            type ArgumentType = Option<super::super::Dessert>;
            const NAME: &'static str = "second";
        }
    }
    pub struct fieldWithInput;
    impl ::cynic::schema::Field for fieldWithInput {
        type Type = super::Dessert;
        const NAME: &'static str = "fieldWithInput";
    }
    impl ::cynic::schema::HasField<fieldWithInput> for super::TestStruct {
        type Type = super::Dessert;
    }
    pub mod field_with_input_arguments {
        pub struct input;
        impl ::cynic::schema::HasArgument<input> for super::fieldWithInput {
            type ArgumentType = super::super::AnInputType;
            const NAME: &'static str = "input";
        }
    }
    pub struct nested;
    impl ::cynic::schema::Field for nested {
        type Type = super::Nested;
        const NAME: &'static str = "nested";
    }
    impl ::cynic::schema::HasField<nested> for super::TestStruct {
        type Type = super::Nested;
    }
    pub struct optNested;
    impl ::cynic::schema::Field for optNested {
        type Type = Option<super::Nested>;
        const NAME: &'static str = "optNested";
    }
    impl ::cynic::schema::HasField<optNested> for super::TestStruct {
        type Type = Option<super::Nested>;
    }
    pub struct dessert;
    impl ::cynic::schema::Field for dessert {
        type Type = Option<super::Dessert>;
        const NAME: &'static str = "dessert";
    }
    impl ::cynic::schema::HasField<dessert> for super::TestStruct {
        type Type = Option<super::Dessert>;
    }
    pub struct json;
    impl ::cynic::schema::Field for json {
        type Type = Option<super::JSON>;
        const NAME: &'static str = "json";
    }
    impl ::cynic::schema::HasField<json> for super::TestStruct {
        type Type = Option<super::JSON>;
    }
}
impl ::cynic::schema::HasSubtype<Nested> for MyUnionType {}
impl ::cynic::schema::HasSubtype<TestStruct> for MyUnionType {}
impl ::cynic::schema::NamedType for MyUnionType {
    const NAME: &'static str = "MyUnionType";
}
impl ::cynic::schema::NamedType for Nested {
    const NAME: &'static str = "Nested";
}
impl ::cynic::schema::NamedType for Query {
    const NAME: &'static str = "Query";
}
impl ::cynic::schema::NamedType for TestStruct {
    const NAME: &'static str = "TestStruct";
}
pub type Boolean = bool;
pub type String = std::string::String;
pub type Float = f64;
pub type Int = i32;
pub type ID = ::cynic::Id;
pub mod variable {
    use cynic::variables::VariableType;
    #[doc = r" Used to determine the type of a given variable that"]
    #[doc = r" appears in an argument struct."]
    pub trait Variable {
        const TYPE: VariableType;
    }
    impl<T> Variable for &T
    where
        T: Variable,
    {
        const TYPE: VariableType = T::TYPE;
    }
    impl<T> Variable for Option<T>
    where
        T: Variable,
    {
        const TYPE: VariableType = VariableType::Nullable(&T::TYPE);
    }
    impl<T> Variable for Vec<T>
    where
        T: Variable,
    {
        const TYPE: VariableType = VariableType::List(&T::TYPE);
    }
    impl<T> Variable for Box<T>
    where
        T: Variable,
    {
        const TYPE: VariableType = T::TYPE;
    }
    impl<T> Variable for std::rc::Rc<T>
    where
        T: Variable,
    {
        const TYPE: VariableType = T::TYPE;
    }
    impl<T> Variable for std::sync::Arc<T>
    where
        T: Variable,
    {
        const TYPE: VariableType = T::TYPE;
    }
    impl Variable for bool {
        const TYPE: VariableType = VariableType::Named("Boolean");
    }
    impl Variable for String {
        const TYPE: VariableType = VariableType::Named("String");
    }
    impl Variable for f64 {
        const TYPE: VariableType = VariableType::Named("Float");
    }
    impl Variable for i32 {
        const TYPE: VariableType = VariableType::Named("Int");
    }
    impl Variable for ::cynic::Id {
        const TYPE: VariableType = VariableType::Named("ID");
    }
}

